### ISSUE TITLE
[NullableReferenceTypes] Track nested slot states in assignments and compound assignments

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#32501]
```C#
class C
{
    object? F = null;
    void M(C c1, C c2)
    {
        if (c1.F == null) return;
        if (c2.F == null) return;

        (c1 = c2).F.ToString(); // Currently a null reference warning
		(c1 ??= c2).F.ToString(); // Currently 
    }
}

```

We should support these scenarios. `VisitAssignmentOperator` and `VisitNullCoalescingAssignmentOperator` should be copying state.

**Comment by jaredpar** [comment]
Related to #30995

**Comment by jcouv** [comment]
The behavior has changed since the bug was filed. There is now a warning on `(c1 ??= c2).F` but not `(c1 = c2).F`.
The behavior seems fine. We're basically saying that `x ??= y` always produces a nullable result.

@333fred Can you add verification for this issue as you update the `??=` operator to return a non-nullable result if the RHS is non-nullable?

**Comment by 333fred** [comment]
@jcouv why would this behavior be fine? That doens't sound fine at all. It's a spurious warning we can suppress.

**Comment by jcouv** [comment]
Hum, you're right. We know that `c2` is non-null, so the result of the compound assignment is also non-null.

Note `(maybeNullString ?? nonNullString).ToString();` also produces a spurious warning.

**Comment by 333fred** [comment]
This has been fixed for regular assignment, but not for null-coalescing assignment.
